_posts/2015-12-09-docker-for-development.markdown

```diff
@@ -0,0 +1,214 @@
+For development which involves multiple applications in separate git repositories, we use [Dusty](http://dusty.gc.com/), which was created by [GameChanger](https://gc.com/). Some of the advantages of using Dusty include the use of `rsync` so that performance is better, and a built-in nginx proxy along with modifications to your `/etc/hosts` file so that you can more easily connect to your applications. The shared volumes approach with the `volumes:` directive can be rather slow with the default Docker Toolbox setup using VirtualBox.
+
+With Dusty, you set up services, apps, and bundles of apps with YAML files. Our MongoDB service is defined as:
+
+```yaml
+# services/mongo2.yml
+image: mongo:2.4
+volumes:
+- /persist/persistentMongo:/data/db
+entrypoint: ["sh", "-c", "rm -f /data/db/mongod.lock; mongod --smallfiles --quiet --logpath=/dev/null"]
+ports:
+- "27017:27017"
+```
+
+It's not necessary to expose the ports, but in case we want to connect directly to the MongoDB instance with the `mongo` command without shelling into a container, we need it to be available on our Docker VM's IP address.
+
+Our Gravity app's YAML file is:
+
+```yaml
+# apps/gravity.yml
+repo: github.com/artsy/gravity
+mount: /app
+build: .
+
+depends:
+  services:
+  - mongo2
+  - memcached
+  - solr4
+  - es15
+  - mailcatcher
+  - redis
+  apps:
+  - radiation
+
+host_forwarding:
+- host_name: gravity
+  host_port: 80
+  container_port: 80
+
+compose:
+  environment:
+    RADIATION_URL: http://radiation
+    MONGO_HOST: mongo2
+    MEMCACHE_SERVERS: memcached
+    SOLR_URL: http://solr4:8983/solr/gravity
+    ELASTICSEARCH_URL: es15:9200
+    SMTP_ADDRESS: mailcatcher
+    SMTP_PORT: 1025
+    REDIS_URL: redis://redis
+
+commands:
+  once:
+  - bundle install -j 10
+  - bundle exec rake db:client_applications:create_all
+  - bundle exec rake db:admin:create
+  always:
+  - rails s -b 0.0.0.0 -p 80
+```
+
+For now, Dusty doesn't have a way of sharing common setup like `common.yml` above, so there are similar configurations for our Sidekiq and Delayed Job workers.
+
+Dusty uses bundles for clusters of applications that need to work together. An example bundle, for a CMS application that needs many APIs, is:
+
+```yaml
+# apps/volt.yml
+description: Volt
+apps:
+  - tangentApi
+  - radiation
+  - superposition
+  - gravity
+  - volt
+```
+
+We bring up that cluster of applications with
+
+```bash
+dusty bundles activate volt
+dusty up
+```
+
+As we have added new services over time, using Docker and Dusty to bring clusters of apps together has made it much easier for developers to work on projects without having to spend time on installations and configuration. Having configuration in code means that we need less documentation, and fewer documentation updates, for how a given application is configured and its dependencies.
+
+We are also starting to use Docker to run integrated testing across multiple applications using Selenium. That will be covered in a future blog post.
```

Comment:
Hook baited, edge of my seat, cannot wait!

![target](https://cloud.githubusercontent.com/assets/197336/11667398/7dfb1352-9dbf-11e5-8574-17ab28b27d73.gif)